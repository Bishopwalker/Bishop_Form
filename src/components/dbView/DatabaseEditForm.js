import React, {useState} from 'react';
import {Link, useLocation} from 'react-router-dom';
import axios from "axios";
import './DatabaseEditForm.css';

const DatabaseEditForm=()=>{
    const location = useLocation();
    const {from} = location.state || {}
    const [firstname,setFirstName] = useState(''),
        [lastname,setLastName] = useState(''),
        [person_age,setAge] = useState(''),
        [hobbies,setHobbies] = useState(''),
        [id,setID] = useState(0),
        [people,setPerson] = useState({
            firstname,
            lastname,
            person_age,
            hobbies
        }),
        [isEdit,setIsEdit] = useState(false);

    const updatePerson= async(id,people)=>{
        await setIsEdit(true);
        try {
            const result = await axios.put(`http://localhost:3003/form/users/${id}`,people)

        } catch (err) {
            console.log(err)
        }
    }
    const handleSubmit=(e)=>{
        e.preventDefault();
        const person = {
            firstname,
            lastname ,
            person_age,
            hobbies
        };
        // console.log(person)
        updatePerson(id,person).then((r)=>setTimeout(()=>setIsEdit(false),3000) )

    }


    function handleFirstNameChange(e) {
        e.preventDefault();
        setFirstName(e.target.value)
    }
    function handleLastNameChange(e) {
        e.preventDefault();
        setLastName(e.target.value)
    }
    function handleAgeChange(e) {
        e.preventDefault();
        setAge(e.target.value)
    }
    function handleHobbiesChange(e) {
        e.preventDefault();
        setHobbies(e.target.value)
    }
    const editPerson = async(id) => {
        try {
            const result = await axios.get(`http://localhost:3003/form/users/${id}`)

            setFirstName(result.data[0].firstname)
            setLastName(result.data[0].lastname)
            setAge(result.data[0].person_age)
            setHobbies(result.data[0].hobbies)
            setID(result.data[0].id)

        } catch (err) {
            console.log(err)
        }

    }
    React.useEffect(()=>{

        editPerson(from).then((r)=>console.log('I promise to edit'))
    },[])
React.useEffect(()=>{
    setPerson({
        firstname,
        lastname,
        person_age,
        hobbies
    })
},[firstname,lastname,person_age,hobbies])

    return(
        <form onSubmit={handleSubmit} data-testid='helper'>

            <input
                type='text'
                placeholder='First Name'
                value={firstname?firstname:''}
                onChange={handleFirstNameChange}
                
            />

            <input
                type='text'
                placeholder='Last Name'
                value={lastname?lastname:''}
                onChange={handleLastNameChange}
                name='lastname'
            />

            <input
                type='number'
                placeholder='Age'
                value={person_age?person_age:''}
                onChange={handleAgeChange}
                name='person_age'
            />

            <input
                type='text'
                placeholder='Hobbies'
                value={hobbies?hobbies:''}
                onChange={handleHobbiesChange}
                name='hobbies'
            />
            <button>Update User</button>
            <Link to='/viewDatabase'>
                <button>return to db</button>
            </Link>
            {isEdit?
                <div>
                    <h4>id: {id}</h4>
                    <h2>name: {people.firstname}, {people.lastname}</h2>
                    <h2>age: {people.person_age}</h2>
                    <h3>hobbies: {people.hobbies}</h3>
                </div>
                :
                <div>
                    <h1>Loading...</h1>
                </div>
            }
        </form>
    )
}
export default DatabaseEditForm;