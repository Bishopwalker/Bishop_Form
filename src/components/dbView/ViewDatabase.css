.alert{
    background-color: mediumvioletred;
    height: 300px;
    font-size: 40px;
}
.alert1{
    background-color: mediumseagreen;
    height: 100%;
}
.alert2{
    background-color: burlywood;
    height: 100%;
}
