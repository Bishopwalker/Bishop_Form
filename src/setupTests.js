// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import {render as rtlRender} from '@testing-library/react';
import {MemoryRouter} from 'react-router-dom'
import {configureStore} from '@reduxjs/toolkit';
import {Provider} from 'react-redux';
import reducer from './ducks/reducer'
import {BrowserRouter} from "react-router-dom";
import {createMemoryHistory} from "history";
//create a customRender that wraps the UI in a memory Router
// const customRender=(ui,options)=>{
//     return render(ui, {wrapper: MemoryRouter, ...options})
// }
// //re-export everything
// export * from '@testing-library/react'
//
// //override render method
//
// export {customRender as render};

function render(
    ui,
    {   route = '/',
        history = createMemoryHistory({initialEntries: [route]}),
        preloadedState,
        store = configureStore({ reducer: { ...reducer, ...preloadedState } }),
        ...renderOptions
    } ={}
) {
    function Wrapper({ children }) {
        return <Provider store={store}>{children}</Provider>;
    }
    return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}
// re-export everything
export * from '@testing-library/react';

// override render method
export {render};
