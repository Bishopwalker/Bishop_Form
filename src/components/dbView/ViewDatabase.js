import React from 'react';
import axios from 'axios';
import "../../App.css"
import {Link} from "react-router-dom"
import './ViewDatabase.css'

const ViewDatabase = (props) => {
  const [data, setData] = React.useState([]);
  const [loading, setLoading] = React.useState(true);
  const [error, setError] = React.useState(null);
  const [ghosting, setGhosting] = React.useState('alert');

React.useEffect(() => {

},[error, loading])
    const fetchData = async () => {
        try {
            const result = await axios.get(
                'http://localhost:3003/form/users/'
            );

            setData(result.data);
            setLoading(false);
        } catch (error) {
            setError(error);

            setLoading(false);
        }
    };
const deletePerson = async(id) => {
    console.log(id)
    try {
        const result = await axios.delete(`http://localhost:3003/form/used/${id}`)

        setData(result.data)
    }catch(err) {
    console.log(err)
    setGhosting('alert2')
  }
}
  React.useEffect(() => {


    fetchData().then(r => console.log(r));
  }, []);

  if (loading) {
    return <div className={ghosting}><h2>Loading...</h2></div>;
  }

  if (error) {

         // setGhosting('alert1');
             setTimeout(() => {
            setGhosting('alert2');
          }, 1000);
             setInterval(() => {
            setGhosting('alert');
             },1000)
    return <div className={ghosting}>Error: {error.message}</div>;
  }
console.log(data)
  let mappedData = data.map((item, index) =>(
      <ul key={index}>

        <li >{item.id}</li>
        <li >{item.firstname} {item.lastname}</li>
        <li>{item.person_age}</li>
        <li>{item.hobbies}</li>
          <div className="btn-group">
          <button onClick={()=>deletePerson(item.id)}>Delete</button>
          <button>Edit</button>
              </div>
      </ul>
  ))
  return (
    <div className="App-header outSide ">
      <div>
          <h2>DataBase Entries</h2>
      {mappedData}
          <Link to='/addUser'>
              <button>Return to Form</button>
          </Link>
      </div>
    </div>
  );
};
export default ViewDatabase;