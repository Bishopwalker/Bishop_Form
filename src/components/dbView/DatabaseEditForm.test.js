import React from 'react'
import {render, cleanup, waitForElement, screen} from '../../setupTests';
import '@testing-library/jest-dom/extend-expect'
import axiosMock from 'axios'
import DatabaseEditForm from './DatabaseEditForm'
import {BrowserRouter, MemoryRouter} from 'react-router-dom'
import {Router} from 'react-router-dom'
import {createMemoryHistory} from 'history'
import ViewDatabase from "./ViewDatabase";

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
  useLocation: () => ({
      pathname: 'localhost:3000/viewSingleUser',
      location: {
        state:{
            person: {
                id:0,
                age: 4,
                lastName: 'Smith',
                firstName: 'Jim',
                hobbies: 'Everythign under the sun',
            }
        }
      }
  })
}))

function renderWithRouter(ui,{route = '/', history=createMemoryHistory({initialEntries: [route,{}]}),location={}} = {}) {
    const state = {
        person: {
            id:0,
            age: 4,
            lastName: 'Smith',
            firstName: 'Jim',
            hobbies: 'Everythign under the sun',
        }
    }
    history.push(
        {pathname: '/database/edit/1'},
        {state: state}
    )
    return {
        ...render(<BrowserRouter history={history} location={history.location}>{ui}</BrowserRouter>),
        history,
    }
}

describe('Database EditForm',()=>{
    afterEach(cleanup);
    // it('It fetches and displays data',async()=>{
    //     const history = createMemoryHistory()
    //     const state = {
    //         person: {
    //             id:0,
    //             age: 4,
    //             lastName: 'Smith',
    //             firstName: 'Jim',
    //             hobbies: 'Everythign under the sun',
    //         }
    //     }
    //     history.push({
    //         pathname: '/database/edit/1',
    //         state: state
    //     })
    // const {} = render(
    //     <Router history={history} location={history.location} navigator={history} initialEntries={[{pathname: '/editDatabase', from:{state:' person: {\n' +
    //                 '                        id:0,\n' +
    //                 '                        age: 4,\n' +
    //                 '                        lastName: \'Smith\',\n' +
    //                 '                        firstName: \'Jim\',\n' +
    //                 '                        hobbies: \'Everythign under the sun\',\n' +
    //                 '                    }'} }]} >
    //         <DatabaseEditForm/>
    //     </Router>
    // );
    // })
    // it('trying another method',()=>{
    //     const history = createMemoryHistory()
    //     const state = {
    //         person: {
    //             id:0,
    //             age: 4,
    //             lastName: 'Smith',
    //             firstName: 'Jim',
    //             hobbies: 'Everythign under the sun',
    //         }
    //     }
    //     history.push({
    //         pathname: '/database/edit/1',
    //         state: state
    //     })
    //     let {getByTestId,getByRole}=  render(
    //         <Router history={history}>
    //         <DatabaseEditForm/>
    //         </Router>,{route: '/database/edit/1',state:state}
    //     );
    // })
    test('Should have a first name input field',async()=>{
      //  renderWithRouter(<DatabaseEditForm/>,{route: 'http://localhost:3003/form/users-test/'})
        render(<DatabaseEditForm/>, {wrapper: BrowserRouter, container: document.body, route:'http://localhost:3003/form/users-test/'});
       expect(await screen.findAllByPlaceholderText(/first name/i)).toHaveLength(1);
    })
    test('Should have a last name input field',async()=>{
        render(<DatabaseEditForm/>, {wrapper: BrowserRouter, container: document.body, route:'http://localhost:3003/form/users-test/'});
        expect(await screen.findAllByPlaceholderText(/last name/i)).toHaveLength(1);
    })
    test('Should have an age input field',async()=>{
        render(<DatabaseEditForm/>, {wrapper: BrowserRouter, container: document.body, route:'http://localhost:3003/form/users-test/'});
        expect(await screen.findAllByPlaceholderText(/age/i)).toHaveLength(1);
    })
    test('Should have a hobbies input field',async()=>{
        render(<DatabaseEditForm/>, {wrapper: BrowserRouter, container: document.body, route:'http://localhost:3003/form/users-test/'});
        expect(await screen.findAllByPlaceholderText(/hobbies/i)).toHaveLength(1);
    })
})