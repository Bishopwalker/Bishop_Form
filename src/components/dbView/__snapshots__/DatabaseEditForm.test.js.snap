// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Database EditForm Should take a snapshot 1`] = `
<DocumentFragment>
  <form
    data-testid="helper"
  >
    <input
      placeholder="First Name"
      type="text"
      value=""
    />
    <input
      name="lastname"
      placeholder="Last Name"
      type="text"
      value=""
    />
    <input
      name="person_age"
      placeholder="Age"
      type="number"
      value=""
    />
    <input
      name="hobbies"
      placeholder="Hobbies"
      type="text"
      value=""
    />
    <button>
      Update User
    </button>
    <a
      href="/viewDatabase"
    >
      <button>
        return to db
      </button>
    </a>
    <div>
      <h1>
        Loading...
      </h1>
    </div>
  </form>
</DocumentFragment>
`;
